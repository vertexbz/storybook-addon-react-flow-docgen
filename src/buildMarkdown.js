// @flow
import type { DocgenInfoType, DocgenPropType, DocgenPropsType, DefaultValueType, FlowDocgenPropType, NativeDocgenPropType } from './types';

const escapeCell = (str: string): string => str
    .replace(/\n/g, ' ')
    .replace(/\|/g, '\\|');

const escape = (str: string): string => escapeCell(str)
    .replace(/_/g, '&lowbar;')
    .replace(/\[/g, '&lsqb;')
    .replace(/]/g, '&rsqb;')
    .replace(/>/g, '&gt;')
    .replace(/</g, '&lt;')
    .replace(/\*/g, '&ast;');

const buildValue = (value: DefaultValueType): string => '`' + escapeCell(String(value.value)) + '`';

const buildNativeType = (type: NativeDocgenPropType): string => {
    if (Array.isArray(type.value)) {
        const stringType = type.value
            .map((typeValue: *): string => typeValue.name || String(typeValue.value))
            .join(' | ');

        return escape(stringType);
    } else if (type.name) {
        return escape(type.name);
    }

    return '';
};

const buildFlowType = (type: FlowDocgenPropType): string => escape(type.raw || type.name);

const buildProp = ([name, prop]: [string, DocgenPropType]): string => '| ' + [
    /* Name        */ '**' + name + '**',
    /* Type        */ prop.flowType ? buildFlowType(prop.flowType) : (prop.type ? buildNativeType(prop.type) : ''),
    /* Required    */ prop.required ? 'yes' : 'no',
    /* Default     */ prop.defaultValue ? buildValue(prop.defaultValue) : '-',
    /* Description */ prop.description  ? escapeCell(prop.description) : ''
].join(' | ') + ' |';

const buildProps = (props: DocgenPropsType): string => `
| Name | Type | Required | Default | Description |
|------|------|:--------:|---------|-------------|
${Object.entries(props).map((buildProp: any)).join('\n')}`;

export default (docgenInfo: DocgenInfoType): string => `
# ${docgenInfo.displayName || 'Unknown'}
${docgenInfo.description || ''}
${docgenInfo.props ? buildProps(docgenInfo.props) : ''}`;
