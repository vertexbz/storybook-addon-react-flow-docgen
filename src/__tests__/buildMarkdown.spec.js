import buildMarkdown from '../buildMarkdown';

describe('buildMarkdown', () => {
    it('builds correct markdown from flow type defs', () => {
        const docgenInfo = {
            'description': 'DemoFlow component description\nblah blah blah\n## more bla',
            'displayName': 'DemoFlow',
            'props': {
                'prop1': {
                    'required': false,
                    'flowType': { 'name': 'number' },
                    'description': '~required~ optional prop1 of type number',
                    'defaultValue': { 'value': '5', 'computed': false }
                },
                'prop2': {
                    'required': false,
                    'flowType': {
                        'name': 'union',
                        'raw': 'string | false',
                        'elements': [{ 'name': 'string' }, { 'name': 'literal', 'value': 'false' }]
                    },
                    'description': 'prop2, required string || `false`',
                    'defaultValue': { 'value': '\'*<asd>*\'', 'computed': false }
                },
                'prop3': {
                    'required': true,
                    'flowType': {
                        'name': 'signature',
                        'type': 'object',
                        'raw': '{\n    a: string,\n    b: number,\n    c?: ?boolean\n}',
                        'signature': {
                            'properties': [{ 'key': 'a', 'value': { 'name': 'string', 'required': true } }, {
                                'key': 'b',
                                'value': { 'name': 'number', 'required': true }
                            }, { 'key': 'c', 'value': { 'name': 'boolean', 'nullable': true, 'required': false } }]
                        }
                    },
                    'description': 'prop3\nrequired **object**'
                },
                'prop4': {
                    'required': true,
                    'flowType': {
                        'name': 'Array',
                        'elements': [{
                            'name': 'signature',
                            'type': 'object',
                            'raw': '{ var: string }',
                            'signature': { 'properties': [{ 'key': 'var', 'value': { 'name': 'string', 'required': true } }] }
                        }],
                        'raw': 'Array<{ var: string }>'
                    },
                    'description': 'prop4, required array of objects'
                },
                'prop5': {
                    'required': true,
                    'flowType': {
                        'name': 'signature',
                        'type': 'function',
                        'raw': '(Array<*>) => void',
                        'signature': {
                            'arguments': [{
                                'name': '',
                                'type': { 'name': 'Array', 'elements': [{ 'name': 'unknown' }], 'raw': 'Array<*>' }
                            }], 'return': { 'name': 'void' }
                        }
                    },
                    'description': 'a *callback*'
                }
            }
        };

        expect(buildMarkdown(docgenInfo)).toMatchSnapshot();
    });
});
