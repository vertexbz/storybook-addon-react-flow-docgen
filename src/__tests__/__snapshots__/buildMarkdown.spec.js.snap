// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMarkdown builds correct markdown from flow type defs 1`] = `
"
# DemoFlow
DemoFlow component description
blah blah blah
## more bla

| Name | Type | Required | Default | Description |
|------|------|:--------:|---------|-------------|
| **prop1** | number | no | \`5\` | ~required~ optional prop1 of type number |
| **prop2** | string \\\\| false | no | \`'*<asd>*'\` | prop2, required string \\\\|\\\\| \`false\` |
| **prop3** | {     a: string,     b: number,     c?: ?boolean } | yes | - | prop3 required **object** |
| **prop4** | Array&lt;{ var: string }&gt; | yes | - | prop4, required array of objects |
| **prop5** | (Array&lt;&ast;&gt;) =&gt; void | yes | - | a *callback* |"
`;
